test/client/dio_interceptor_test.dart

```diff
@@ -0,0 +1,344 @@
+          isDateInHeader: true,
+          isToleranceHighEnough: false),
+      TestCase(
+          isAutoRetryEnabled: true,
+          isDateInHeader: false,
+          isToleranceHighEnough: true),
+      TestCase(
+          isAutoRetryEnabled: true,
+          isDateInHeader: false,
+          isToleranceHighEnough: false),
+      TestCase(
+          isAutoRetryEnabled: false,
+          isDateInHeader: true,
+          isToleranceHighEnough: true),
+      TestCase(
+          isAutoRetryEnabled: false,
+          isDateInHeader: true,
+          isToleranceHighEnough: false),
+      TestCase(
+          isAutoRetryEnabled: false,
+          isDateInHeader: false,
+          isToleranceHighEnough: true),
+      TestCase(
+          isAutoRetryEnabled: false,
+          isDateInHeader: false,
+          isToleranceHighEnough: false),
+    ];
+
+    httpStatusCoveredByAutoRetry.forEach((serverHttpStatus) {
+      test(
+          "Interceptor should autoretry on $serverHttpStatus received when it's enabled, a date is set in the response header and the tolerence is high enough.",
+          () async {
+        // Arrange
+        final options = RequestsSignatureOptions(
+          clientId: testClientId,
+          clientSecret: testClientSecret,
+          headerName: headerName,
+          signaturePattern: signaturePattern,
+          disableAutoRetryOnClockSkew: false,
+          clockSkew: Duration(milliseconds: clockskewMS),
+        );
+
+        final dioAdapter = DioAdapter(dio: dio);
+        dio.httpClientAdapter = dioAdapter;
+
+        // Create the interceptor
+        final interceptor = RequestsSignatureInterceptor(
+          options,
+          dio,
+        );
+
+        // Add the interceptor to Dio
+        dio.interceptors.add(interceptor);
+        var mockServer = MockServerValidator(
+          headerName,
+          clockskewMS,
+          signaturePattern,
+        );
+
+        var callCount = 0;
+
+        // Register the mock response
+        dioAdapter.onGet(
+          baseEndpoint,
+          (server) {
+            Map<String, List<String>> headers = {
+              "date": [DateTime.now().toString()]
+            };
+
+            server.replyCallbackAsync(serverHttpStatus, (request) async {
+              callCount++;
+
+              await Future.delayed(Duration(milliseconds: clockskewMS));
+
+              var isValid =
+                  await mockServer.validateSignature(request, _uuid.v4());
+
+              dioAdapter.onGet(
+                baseEndpoint,
+                (server) {
+                  callCount++;
+                  Map<String, List<String>> headers = {};
+
+                  server.reply(
+                      serverHttpStatus, (request) => emptyDataInjection,
+                      headers: headers);
+                },
+              );
+
+              if (isValid) {
+                throw new Exception("The signature should not be valid.");
```

Comment:
throw a friendlier exception

Reply:
I think I made it friendlier. I've also added a comment that explains why we're throwing exceptions here